cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```
@@ -614,7 +614,7 @@ private LocalFile uploadFile(String bucket, String targetFilePath,
         Blob existingBlob = storage.get(blobId);
         BlobInfo newBlobInfo = existingBlob.toBuilder().setContentType(contentType).build();
         uploadToGcsUtil(localFile, storage, targetFilePath, newBlobInfo,
-            Storage.BlobWriteOption.generationNotMatch());
+            Storage.BlobWriteOption.generationMatch());
```

What would happen if generationNotMatch is the case when we reach here?

Why not to ignore the exception if generationMatch precondition holds?

---

This logic is becoming a bit complex. I believe we need to add couples of unittests to ensure we correctly cover all cases.

---

> What would happen if generationNotMatch is the case when we reach here?
> 
> Why not to ignore the exception if generationMatch precondition holds?

When we are uploading the new file we do not have a generation associated with the blob i.e generation=0.

```
The generation number is a monotonically increasing integer that updates when the data associated with an object name changes. 
Thus an object name and generation number uniquely identify a specific data resource that can be deleted, but never replaced with different data
```

The pre-condition can fail in the next attempt that means the data in the object has changed which can be due to partial/complete upload in the previous attempt and new generation number will be attached to the object. Hence, getting the blob information and retrying the operation.

---

Where do we retry?

---

The retry strategy is configured in the storage client: https://github.com/cdapio/cdap/blob/0366675489336099dd8dcbbd5a96a8e0a5159afa/cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java#L197-L206

---

Can you add comments regarding why match precondition is being used.

---

I've removed the precondition from second attempt, since we are using `UniformRetryStrategy` which also retries all operations.